File: code/packages/elixir/state_machine/test/nfa_test.exs
defmodule CodingAdventures.StateMachine.NFATest do
  use ExUnit.Case, async: true

  alias CodingAdventures.StateMachine.NFA
  alias CodingAdventures.StateMachine.DFA

  # === Fixture helpers ===

  # NFA that accepts strings containing the substring "ab".
  #
  # This is non-deterministic because in q0, on input "a", the machine
  # can either stay in q0 (guessing 'ab' hasn't started) or move to q1
  # (guessing this 'a' starts the 'ab' pattern).
  defp contains_ab do
    NFA.new(
      MapSet.new(["q0", "q1", "q2"]),
      MapSet.new(["a", "b"]),
      %{
        {"q0", "a"} => MapSet.new(["q0", "q1"]),
        {"q0", "b"} => MapSet.new(["q0"]),
        {"q1", "b"} => MapSet.new(["q2"]),
        {"q2", "a"} => MapSet.new(["q2"]),
        {"q2", "b"} => MapSet.new(["q2"])
      },
      "q0",
      MapSet.new(["q2"])
    )
  end

  # NFA with epsilon transitions: accepts "a" or "b" (union).
  #
  # q0 --epsilon--> q1 --a--> q3 (accept)
  # q0 --epsilon--> q2 --b--> q3 (accept)
  defp epsilon_union do
    NFA.new(
      MapSet.new(["q0", "q1", "q2", "q3"]),
      MapSet.new(["a", "b"]),
      %{
        {"q0", ""} => MapSet.new(["q1", "q2"]),
        {"q1", "a"} => MapSet.new(["q3"]),
        {"q2", "b"} => MapSet.new(["q3"])
      },
      "q0",
      MapSet.new(["q3"])
    )
  end

  # Simple NFA that accepts "a" then "b" with no non-determinism.
  # Used to test basic functionality.
  defp simple_ab do
    NFA.new(
      MapSet.new(["q0", "q1", "q2"]),
      MapSet.new(["a", "b"]),
      %{
        {"q0", "a"} => MapSet.new(["q1"]),
        {"q1", "b"} => MapSet.new(["q2"])
      },
      "q0",
      MapSet.new(["q2"])
    )
  end

  # NFA with chained epsilon transitions: q0 --eps--> q1 --eps--> q2 --a--> q3.
  defp chained_epsilon do
    NFA.new(
      MapSet.new(["q0", "q1", "q2", "q3"]),
      MapSet.new(["a"]),
      %{
        {"q0", ""} => MapSet.new(["q1"]),
        {"q1", ""} => MapSet.new(["q2"]),
        {"q2", "a"} => MapSet.new(["q3"])
      },
      "q0",
      MapSet.new(["q3"])
    )
  end

  # === Construction tests ===

  describe "new/5 — construction and validation" do
    test "creates a valid NFA" do
      assert {:ok, nfa} = contains_ab()
      assert nfa.initial == "q0"
      assert MapSet.member?(nfa.current, "q0")
    end

    test "initial current includes epsilon closure" do
      {:ok, nfa} = epsilon_union()
      # q0 has epsilon transitions to q1 and q2
      assert MapSet.member?(nfa.current, "q0")
      assert MapSet.member?(nfa.current, "q1")
      assert MapSet.member?(nfa.current, "q2")
    end

    test "rejects empty states" do
      assert {:error, msg} =
               NFA.new(MapSet.new(), MapSet.new(["a"]), %{}, "q0", MapSet.new())

      assert msg =~ "non-empty"
    end

    test "rejects epsilon in alphabet" do
      assert {:error, msg} =
               NFA.new(
                 MapSet.new(["q0"]),
                 MapSet.new(["a", ""]),
                 %{},
                 "q0",
                 MapSet.new()
               )

      assert msg =~ "epsilon"
    end

    test "rejects initial state not in states" do
      assert {:error, msg} =
               NFA.new(
                 MapSet.new(["q0"]),
                 MapSet.new(["a"]),
                 %{},
                 "q99",
                 MapSet.new()
               )

      assert msg =~ "q99"
    end

    test "rejects accepting states not in states" do
      assert {:error, msg} =
               NFA.new(
                 MapSet.new(["q0"]),
                 MapSet.new(["a"]),
                 %{},
                 "q0",
                 MapSet.new(["q99"])
               )

      assert msg =~ "q99"
    end

    test "rejects transition with unknown source" do
      assert {:error, msg} =
               NFA.new(
                 MapSet.new(["q0"]),
                 MapSet.new(["a"]),
                 %{{"q99", "a"} => MapSet.new(["q0"])},
                 "q0",
                 MapSet.new()
               )

      assert msg =~ "q99"
    end

    test "rejects transition with unknown event" do
      assert {:error, msg} =
               NFA.new(
                 MapSet.new(["q0"]),
                 MapSet.new(["a"]),
                 %{{"q0", "z"} => MapSet.new(["q0"])},
                 "q0",
                 MapSet.new()
               )

      assert msg =~ "z"
    end

    test "rejects transition with unknown target" do
      assert {:error, msg} =
               NFA.new(
                 MapSet.new(["q0"]),
                 MapSet.new(["a"]),
                 %{{"q0", "a"} => MapSet.new(["q99"])},
                 "q0",
                 MapSet.new()
               )

      assert msg =~ "q99"
    end
  end

  # === Epsilon closure tests ===

  describe "epsilon_closure/2" do
    test "closure of state with no epsilon transitions is itself" do
      {:ok, nfa} = contains_ab()
      closure = NFA.epsilon_closure(nfa, MapSet.new(["q0"]))
      assert closure == MapSet.new(["q0"])
    end

    test "follows single epsilon transition" do
      {:ok, nfa} = epsilon_union()
      closure = NFA.epsilon_closure(nfa, MapSet.new(["q0"]))
      assert MapSet.equal?(closure, MapSet.new(["q0", "q1", "q2"]))
    end

    test "follows chained epsilon transitions" do
      {:ok, nfa} = chained_epsilon()
      closure = NFA.epsilon_closure(nfa, MapSet.new(["q0"]))
      assert MapSet.equal?(closure, MapSet.new(["q0", "q1", "q2"]))
    end

    test "closure of empty set is empty" do
      {:ok, nfa} = contains_ab()
      assert NFA.epsilon_closure(nfa, MapSet.new()) == MapSet.new()
    end

    test "closure of multiple states is union" do
      {:ok, nfa} = epsilon_union()
      closure = NFA.epsilon_closure(nfa, MapSet.new(["q1", "q2"]))
      assert MapSet.equal?(closure, MapSet.new(["q1", "q2"]))
    end
  end

  # === Processing tests ===

  describe "process/2" do
    test "processes single event" do
      {:ok, nfa} = contains_ab()
      {:ok, nfa} = NFA.process(nfa, "a")
      # In q0, "a" goes to {q0, q1}
      assert MapSet.member?(nfa.current, "q0")
      assert MapSet.member?(nfa.current, "q1")
    end

    test "non-deterministic branching" do
      {:ok, nfa} = contains_ab()
      {:ok, nfa} = NFA.process(nfa, "a")
      {:ok, nfa} = NFA.process(nfa, "b")
      # q0 on b -> q0, q1 on b -> q2
      assert MapSet.member?(nfa.current, "q0")
      assert MapSet.member?(nfa.current, "q2")
    end

    test "dead branches are pruned" do
      {:ok, nfa} = simple_ab()
      {:ok, nfa} = NFA.process(nfa, "a")
      {:ok, nfa} = NFA.process(nfa, "a")
      # q1 has no transition on "a", only q0->a transitions work
      # But there is no transition for q1 on "a", so that branch dies
      # and q0 has no transition on "a" either (only q0->a->q1)
      # Wait — q0 on "a" -> q1, q1 on "a" -> nothing
      # After first "a": current = {q1}
      # After second "a": current = {} (dead)
      assert MapSet.size(nfa.current) == 0
    end

    test "rejects unknown event" do
      {:ok, nfa} = contains_ab()
      assert {:error, msg} = NFA.process(nfa, "z")
      assert msg =~ "z"
    end

    test "epsilon transitions applied after event" do
      {:ok, nfa} = epsilon_union()
      {:ok, nfa} = NFA.process(nfa, "a")
      assert MapSet.member?(nfa.current, "q3")
    end
  end

  # === Acceptance tests ===

  describe "accepts?/2" do
    test "contains_ab accepts 'ab'" do
      {:ok, nfa} = contains_ab()
      assert NFA.accepts?(nfa, ["a", "b"])
    end

    test "contains_ab accepts 'aab'" do
      {:ok, nfa} = contains_ab()
      assert NFA.accepts?(nfa, ["a", "a", "b"])
    end

    test "contains_ab accepts 'abb'" do
      {:ok, nfa} = contains_ab()
      assert NFA.accepts?(nfa, ["a", "b", "b"])
    end

    test "contains_ab accepts 'bab'" do
      {:ok, nfa} = contains_ab()
      assert NFA.accepts?(nfa, ["b", "a", "b"])
    end

    test "contains_ab rejects 'ba'" do
      {:ok, nfa} = contains_ab()
      refute NFA.accepts?(nfa, ["b", "a"])
    end

    test "contains_ab rejects 'a'" do
      {:ok, nfa} = contains_ab()
      refute NFA.accepts?(nfa, ["a"])
    end

    test "contains_ab rejects empty" do
      {:ok, nfa} = contains_ab()
      refute NFA.accepts?(nfa, [])
    end

    test "epsilon_union accepts 'a'" do
      {:ok, nfa} = epsilon_union()
      assert NFA.accepts?(nfa, ["a"])
    end

    test "epsilon_union accepts 'b'" do
      {:ok, nfa} = epsilon_union()
      assert NFA.accepts?(nfa, ["b"])
    end

    test "epsilon_union rejects 'ab'" do
      {:ok, nfa} = epsilon_union()
      refute NFA.accepts?(nfa, ["a", "b"])
    end

    test "epsilon_union rejects empty" do
      {:ok, nfa} = epsilon_union()
      refute NFA.accepts?(nfa, [])
    end

    test "chained epsilon NFA accepts 'a'" do
      {:ok, nfa} = chained_epsilon()
      assert NFA.accepts?(nfa, ["a"])
    end

    test "chained epsilon NFA rejects empty" do
      {:ok, nfa} = chained_epsilon()
      refute NFA.accepts?(nfa, [])
    end

    test "accepts? does not modify the NFA" do
      {:ok, nfa} = contains_ab()
      original_current = nfa.current
      NFA.accepts?(nfa, ["a", "b"])
      assert nfa.current == original_current
    end

    test "simple_ab accepts exactly 'ab'" do
      {:ok, nfa} = simple_ab()
      assert NFA.accepts?(nfa, ["a", "b"])
      refute NFA.accepts?(nfa, ["a"])
      refute NFA.accepts?(nfa, ["b"])
      refute NFA.accepts?(nfa, ["b", "a"])
      refute NFA.accepts?(nfa, ["a", "b", "a"])
    end

    test "raises on invalid event" do
      {:ok, nfa} = contains_ab()

      assert_raise ArgumentError, fn ->
        NFA.accepts?(nfa, ["z"])
      end
    end
  end

  # === Reset tests ===

  describe "reset/1" do
    test "resets to initial epsilon closure" do
      {:ok, nfa} = epsilon_union()
      {:ok, nfa} = NFA.process(nfa, "a")
      nfa = NFA.reset(nfa)

      assert MapSet.member?(nfa.current, "q0")
      assert MapSet.member?(nfa.current, "q1")
      assert MapSet.member?(nfa.current, "q2")
    end

    test "resets after processing" do
      {:ok, nfa} = contains_ab()
      {:ok, nfa} = NFA.process(nfa, "a")
      {:ok, nfa} = NFA.process(nfa, "b")
      nfa = NFA.reset(nfa)
      assert nfa.current == MapSet.new(["q0"])
    end
  end

  # === Conversion to DFA tests ===

  describe "to_dfa/1" do
    test "converted DFA accepts same language as NFA — contains_ab" do
      {:ok, nfa} = contains_ab()
      {:ok, dfa} = NFA.to_dfa(nfa)

      # Test several strings
      assert DFA.accepts?(dfa, ["a", "b"])
      assert DFA.accepts?(dfa, ["a", "a", "b"])
      assert DFA.accepts?(dfa, ["b", "a", "b"])
      refute DFA.accepts?(dfa, ["b", "a"])
      refute DFA.accepts?(dfa, ["a"])
      refute DFA.accepts?(dfa, [])
    end

    test "converted DFA accepts same language — epsilon_union" do
      {:ok, nfa} = epsilon_union()
      {:ok, dfa} = NFA.to_dfa(nfa)

      assert DFA.accepts?(dfa, ["a"])
      assert DFA.accepts?(dfa, ["b"])
      refute DFA.accepts?(dfa, ["a", "b"])
      refute DFA.accepts?(dfa, [])
    end

    test "converted DFA accepts same language — chained_epsilon" do
      {:ok, nfa} = chained_epsilon()
      {:ok, dfa} = NFA.to_dfa(nfa)

      assert DFA.accepts?(dfa, ["a"])
      refute DFA.accepts?(dfa, [])
    end

    test "converted DFA has valid structure" do
      {:ok, nfa} = contains_ab()
      {:ok, dfa} = NFA.to_dfa(nfa)

      assert MapSet.size(dfa.states) > 0
      assert MapSet.size(dfa.alphabet) > 0
      assert map_size(dfa.transitions) > 0
    end

    test "DFA state names are deterministic" do
      {:ok, nfa} = contains_ab()
      {:ok, dfa1} = NFA.to_dfa(nfa)
      {:ok, dfa2} = NFA.to_dfa(nfa)

      assert dfa1.states == dfa2.states
      assert dfa1.initial == dfa2.initial
    end
  end

  # === Visualization tests ===

  describe "to_dot/1" do
    test "generates valid DOT" do
      {:ok, nfa} = contains_ab()
      dot = NFA.to_dot(nfa)

      assert dot =~ "digraph NFA"
      assert dot =~ "rankdir=LR"
      assert dot =~ "__start"
    end

    test "marks accepting states" do
      {:ok, nfa} = contains_ab()
      dot = NFA.to_dot(nfa)
      assert dot =~ "doublecircle"
    end

    test "shows epsilon transitions" do
      {:ok, nfa} = epsilon_union()
      dot = NFA.to_dot(nfa)
      # Should contain the epsilon character
      assert dot =~ "\u03b5"
    end
  end

  # === state_set_name tests ===

  describe "state_set_name/1" do
    test "single state" do
      assert NFA.state_set_name(MapSet.new(["q0"])) == "{q0}"
    end

    test "multiple states sorted" do
      assert NFA.state_set_name(MapSet.new(["q2", "q0", "q1"])) == "{q0,q1,q2}"
    end

    test "empty set" do
      assert NFA.state_set_name(MapSet.new()) == "{}"
    end
  end

  # === Edge cases ===

  describe "edge cases" do
    test "NFA with only epsilon transitions" do
      {:ok, nfa} =
        NFA.new(
          MapSet.new(["q0", "q1"]),
          MapSet.new(["a"]),
          %{{"q0", ""} => MapSet.new(["q1"])},
          "q0",
          MapSet.new(["q1"])
        )

      # Should accept empty string because initial epsilon closure reaches q1
      assert NFA.accepts?(nfa, [])
    end

    test "NFA with no transitions" do
      {:ok, nfa} =
        NFA.new(
          MapSet.new(["q0"]),
          MapSet.new(["a"]),
          %{},
          "q0",
          MapSet.new(["q0"])
        )

      assert NFA.accepts?(nfa, [])
      refute NFA.accepts?(nfa, ["a"])
    end

    test "NFA with self-loop epsilon does not infinite loop" do
      # This shouldn't happen in practice, but epsilon_closure must handle it
      {:ok, nfa} =
        NFA.new(
          MapSet.new(["q0"]),
          MapSet.new(["a"]),
          %{{"q0", ""} => MapSet.new(["q0"])},
          "q0",
          MapSet.new(["q0"])
        )

      closure = NFA.epsilon_closure(nfa, MapSet.new(["q0"]))
      assert closure == MapSet.new(["q0"])
    end
  end
end

File: code/packages/go/state-machine/nfa_test.go
package statemachine

import (
	"strings"
	"testing"
)

// =========================================================================
// NFA Test helpers
// =========================================================================

// makeContainsABNFA creates an NFA that accepts strings containing "ab":
//
//	q0 --a--> {q0, q1}  (non-deterministic: stay or start matching)
//	q0 --b--> {q0}
//	q1 --b--> {q2}
//	q2 --a--> {q2}
//	q2 --b--> {q2}
func makeContainsABNFA() *NFA {
	return NewNFA(
		[]string{"q0", "q1", "q2"},
		[]string{"a", "b"},
		map[[2]string][]string{
			{"q0", "a"}: {"q0", "q1"},
			{"q0", "b"}: {"q0"},
			{"q1", "b"}: {"q2"},
			{"q2", "a"}: {"q2"},
			{"q2", "b"}: {"q2"},
		},
		"q0",
		[]string{"q2"},
	)
}

// makeEpsilonNFA creates an NFA with epsilon transitions:
//
//	q0 --epsilon--> q1
//	q1 --a--> q2
//	q2 --epsilon--> q3
func makeEpsilonNFA() *NFA {
	return NewNFA(
		[]string{"q0", "q1", "q2", "q3"},
		[]string{"a"},
		map[[2]string][]string{
			{"q0", EPSILON}: {"q1"},
			{"q1", "a"}:     {"q2"},
			{"q2", EPSILON}: {"q3"},
		},
		"q0",
		[]string{"q3"},
	)
}

// makeMultiEpsilonNFA creates an NFA with chained epsilon transitions:
//
//	q0 --epsilon--> q1
//	q1 --epsilon--> q2
//	q2 --a--> q3
func makeMultiEpsilonNFA() *NFA {
	return NewNFA(
		[]string{"q0", "q1", "q2", "q3"},
		[]string{"a"},
		map[[2]string][]string{
			{"q0", EPSILON}: {"q1"},
			{"q1", EPSILON}: {"q2"},
			{"q2", "a"}:     {"q3"},
		},
		"q0",
		[]string{"q3"},
	)
}

// makeEndsWith01NFA creates an NFA that accepts binary strings ending with "01".
func makeEndsWith01NFA() *NFA {
	return NewNFA(
		[]string{"q0", "q1", "q2"},
		[]string{"0", "1"},
		map[[2]string][]string{
			{"q0", "0"}: {"q0", "q1"},
			{"q0", "1"}: {"q0"},
			{"q1", "1"}: {"q2"},
		},
		"q0",
		[]string{"q2"},
	)
}

// =========================================================================
// Construction tests
// =========================================================================

func TestNewNFA_Valid(t *testing.T) {
	nfa := makeContainsABNFA()

	if nfa.Initial() != "q0" {
		t.Errorf("Initial = %q, want %q", nfa.Initial(), "q0")
	}

	states := nfa.States()
	if len(states) != 3 {
		t.Errorf("len(States) = %d, want 3", len(states))
	}

	alpha := nfa.Alphabet()
	if len(alpha) != 2 {
		t.Errorf("len(Alphabet) = %d, want 2", len(alpha))
	}

	acc := nfa.Accepting()
	if len(acc) != 1 || acc[0] != "q2" {
		t.Errorf("Accepting = %v, want [q2]", acc)
	}
}

func TestNewNFA_EmptyStates(t *testing.T) {
	assertPanics(t, "empty states", "non-empty", func() {
		NewNFA([]string{}, []string{"a"}, nil, "q0", nil)
	})
}

func TestNewNFA_EpsilonInAlphabet(t *testing.T) {
	assertPanics(t, "epsilon in alphabet", "empty string", func() {
		NewNFA([]string{"q0"}, []string{""}, nil, "q0", nil)
	})
}

func TestNewNFA_InvalidInitial(t *testing.T) {
	assertPanics(t, "invalid initial", "initial state", func() {
		NewNFA([]string{"q0"}, []string{"a"}, nil, "bad", nil)
	})
}

func TestNewNFA_InvalidAccepting(t *testing.T) {
	assertPanics(t, "invalid accepting", "accepting state", func() {
		NewNFA([]string{"q0"}, []string{"a"}, nil, "q0", []string{"bad"})
	})
}

func TestNewNFA_InvalidTransitionSource(t *testing.T) {
	assertPanics(t, "invalid source", "transition source", func() {
		NewNFA(
			[]string{"q0"}, []string{"a"},
			map[[2]string][]string{{"bad", "a"}: {"q0"}},
			"q0", nil,
		)
	})
}

func TestNewNFA_InvalidTransitionEvent(t *testing.T) {
	assertPanics(t, "invalid event", "transition event", func() {
		NewNFA(
			[]string{"q0"}, []string{"a"},
			map[[2]string][]string{{"q0", "bad"}: {"q0"}},
			"q0", nil,
		)
	})
}

func TestNewNFA_InvalidTransitionTarget(t *testing.T) {
	assertPanics(t, "invalid target", "transition target", func() {
		NewNFA(
			[]string{"q0"}, []string{"a"},
			map[[2]string][]string{{"q0", "a"}: {"bad"}},
			"q0", nil,
		)
	})
}

// =========================================================================
// Epsilon closure tests
// =========================================================================

func TestNFA_EpsilonClosure_NoEpsilon(t *testing.T) {
	nfa := makeContainsABNFA()
	closure := nfa.EpsilonClosure(map[string]bool{"q0": true})

	if len(closure) != 1 || !closure["q0"] {
		t.Errorf("EpsilonClosure({q0}) = %v, want {q0}", closure)
	}
}

func TestNFA_EpsilonClosure_SingleEpsilon(t *testing.T) {
	nfa := makeEpsilonNFA()
	closure := nfa.EpsilonClosure(map[string]bool{"q0": true})

	if !closure["q0"] || !closure["q1"] {
		t.Errorf("EpsilonClosure({q0}) = %v, want {q0, q1}", closure)
	}
	if len(closure) != 2 {
		t.Errorf("EpsilonClosure size = %d, want 2", len(closure))
	}
}

func TestNFA_EpsilonClosure_Chained(t *testing.T) {
	nfa := makeMultiEpsilonNFA()
	closure := nfa.EpsilonClosure(map[string]bool{"q0": true})

	for _, s := range []string{"q0", "q1", "q2"} {
		if !closure[s] {
			t.Errorf("EpsilonClosure({q0}) should contain %q", s)
		}
	}
	if closure["q3"] {
		t.Error("EpsilonClosure({q0}) should not contain q3 (requires input 'a')")
	}
}

func TestNFA_EpsilonClosure_MultipleStart(t *testing.T) {
	nfa := makeEpsilonNFA()
	closure := nfa.EpsilonClosure(map[string]bool{"q0": true, "q2": true})

	// q0 -> q1 via epsilon; q2 -> q3 via epsilon
	for _, s := range []string{"q0", "q1", "q2", "q3"} {
		if !closure[s] {
			t.Errorf("EpsilonClosure({q0, q2}) should contain %q", s)
		}
	}
}

func TestNFA_EpsilonClosure_Empty(t *testing.T) {
	nfa := makeContainsABNFA()
	closure := nfa.EpsilonClosure(map[string]bool{})
	if len(closure) != 0 {
		t.Errorf("EpsilonClosure({}) = %v, want empty", closure)
	}
}

// =========================================================================
// Initial state includes epsilon closure
// =========================================================================

func TestNFA_InitialEpsilonClosure(t *testing.T) {
	nfa := makeEpsilonNFA()
	current := nfa.CurrentStates()

	if !current["q0"] || !current["q1"] {
		t.Errorf("initial CurrentStates = %v, want {q0, q1}", current)
	}
}

// =========================================================================
// Processing tests
// =========================================================================

func TestNFA_Process(t *testing.T) {
	nfa := makeContainsABNFA()

	// Process "a": from {q0}, on 'a', go to {q0, q1} (no epsilon transitions)
	states := nfa.Process("a")
	if !states["q0"] || !states["q1"] {
		t.Errorf("after 'a': states = %v, want {q0, q1}", states)
	}

	// Process "b": from {q0, q1}, on 'b', go to {q0, q2}
	states = nfa.Process("b")
	if !states["q0"] || !states["q2"] {
		t.Errorf("after 'ab': states = %v, want {q0, q2}", states)
	}
}

func TestNFA_Process_InvalidEvent(t *testing.T) {
	nfa := makeContainsABNFA()
	assertPanics(t, "invalid event", "not in the alphabet", func() {
		nfa.Process("c")
	})
}

func TestNFA_Process_EpsilonTransitions(t *testing.T) {
	nfa := makeEpsilonNFA()
	// Start: {q0, q1} (epsilon closure of q0)
	// Process "a": q1 --a--> q2, then epsilon: q2 --eps--> q3
	states := nfa.Process("a")

	if !states["q2"] || !states["q3"] {
		t.Errorf("after 'a': states = %v, want {q2, q3}", states)
	}
}

// =========================================================================
// Accepts tests
// =========================================================================

func TestNFA_Accepts_ContainsAB(t *testing.T) {
	nfa := makeContainsABNFA()

	tests := []struct {
		name   string
		events []string
		want   bool
	}{
		{"ab", []string{"a", "b"}, true},
		{"aab", []string{"a", "a", "b"}, true},
		{"bab", []string{"b", "a", "b"}, true},
		{"abb", []string{"a", "b", "b"}, true},
		{"a", []string{"a"}, false},
		{"b", []string{"b"}, false},
		{"ba", []string{"b", "a"}, false},
		{"empty", []string{}, false},
		{"bbb", []string{"b", "b", "b"}, false},
		{"abab", []string{"a", "b", "a", "b"}, true},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := nfa.Accepts(tt.events)
			if got != tt.want {
				t.Errorf("Accepts(%v) = %v, want %v", tt.events, got, tt.want)
			}
		})
	}
}

func TestNFA_Accepts_DoesNotMutate(t *testing.T) {
	nfa := makeContainsABNFA()
	nfa.Process("a")
	before := nfa.CurrentStates()

	_ = nfa.Accepts([]string{"b"})

	after := nfa.CurrentStates()
	if len(before) != len(after) {
		t.Error("Accepts mutated the NFA's state")
	}
}

func TestNFA_Accepts_InvalidEvent(t *testing.T) {
	nfa := makeContainsABNFA()
	assertPanics(t, "accepts invalid event", "not in the alphabet", func() {
		nfa.Accepts([]string{"c"})
	})
}

func TestNFA_Accepts_EpsilonNFA(t *testing.T) {
	nfa := makeEpsilonNFA()

	tests := []struct {
		events []string
		want   bool
	}{
		{[]string{"a"}, true},
		{[]string{}, false},
		{[]string{"a", "a"}, false},
	}

	for _, tt := range tests {
		got := nfa.Accepts(tt.events)
		if got != tt.want {
			t.Errorf("epsilon NFA Accepts(%v) = %v, want %v", tt.events, got, tt.want)
		}
	}
}

func TestNFA_Accepts_EndsWith01(t *testing.T) {
	nfa := makeEndsWith01NFA()

	tests := []struct {
		events []string
		want   bool
	}{
		{[]string{"0", "1"}, true},
		{[]string{"1", "0", "1"}, true},
		{[]string{"0", "0", "1"}, true},
		{[]string{"1"}, false},
		{[]string{"0"}, false},
		{[]string{"1", "0"}, false},
		{[]string{}, false},
	}

	for _, tt := range tests {
		got := nfa.Accepts(tt.events)
		if got != tt.want {
			t.Errorf("ends-with-01 NFA Accepts(%v) = %v, want %v", tt.events, got, tt.want)
		}
	}
}

// =========================================================================
// Reset tests
// =========================================================================

func TestNFA_Reset(t *testing.T) {
	nfa := makeContainsABNFA()
	nfa.Process("a")
	nfa.Process("b")
	nfa.Reset()

	current := nfa.CurrentStates()
	if len(current) != 1 || !current["q0"] {
		t.Errorf("after reset: CurrentStates = %v, want {q0}", current)
	}
}

func TestNFA_Reset_EpsilonNFA(t *testing.T) {
	nfa := makeEpsilonNFA()
	nfa.Process("a")
	nfa.Reset()

	current := nfa.CurrentStates()
	if !current["q0"] || !current["q1"] {
		t.Errorf("after reset: CurrentStates = %v, want {q0, q1}", current)
	}
}

// =========================================================================
// ToDFA (subset construction) tests
// =========================================================================

func TestNFA_ToDFA_ContainsAB(t *testing.T) {
	nfa := makeContainsABNFA()
	dfa := nfa.ToDFA()

	// The DFA should accept the same language
	tests := []struct {
		events []string
		want   bool
	}{
		{[]string{"a", "b"}, true},
		{[]string{"a", "a", "b"}, true},
		{[]string{"b", "a", "b"}, true},
		{[]string{"a"}, false},
		{[]string{"b"}, false},
		{[]string{}, false},
	}

	for _, tt := range tests {
		got := dfa.Accepts(tt.events)
		if got != tt.want {
			t.Errorf("DFA from NFA: Accepts(%v) = %v, want %v", tt.events, got, tt.want)
		}
	}
}

func TestNFA_ToDFA_EpsilonNFA(t *testing.T) {
	nfa := makeEpsilonNFA()
	dfa := nfa.ToDFA()

	if !dfa.Accepts([]string{"a"}) {
		t.Error("DFA from epsilon NFA should accept [a]")
	}
	if dfa.Accepts([]string{}) {
		t.Error("DFA from epsilon NFA should reject []")
	}
	if dfa.Accepts([]string{"a", "a"}) {
		t.Error("DFA from epsilon NFA should reject [a, a]")
	}
}

func TestNFA_ToDFA_EndsWith01(t *testing.T) {
	nfa := makeEndsWith01NFA()
	dfa := nfa.ToDFA()

	tests := []struct {
		events []string
		want   bool
	}{
		{[]string{"0", "1"}, true},
		{[]string{"1", "0", "1"}, true},
		{[]string{"0"}, false},
		{[]string{"1", "0"}, false},
	}

	for _, tt := range tests {
		got := dfa.Accepts(tt.events)
		if got != tt.want {
			t.Errorf("DFA from ends-with-01 NFA: Accepts(%v) = %v, want %v", tt.events, got, tt.want)
		}
	}
}

func TestNFA_ToDFA_MultiEpsilon(t *testing.T) {
	nfa := makeMultiEpsilonNFA()
	dfa := nfa.ToDFA()

	if !dfa.Accepts([]string{"a"}) {
		t.Error("DFA from multi-epsilon NFA should accept [a]")
	}
	if dfa.Accepts([]string{}) {
		t.Error("DFA from multi-epsilon NFA should reject []")
	}
}

func TestNFA_ToDFA_StateNames(t *testing.T) {
	nfa := makeContainsABNFA()
	dfa := nfa.ToDFA()

	// DFA state names should be sets in braces
	for _, s := range dfa.States() {
		if !strings.HasPrefix(s, "{") || !strings.HasSuffix(s, "}") {
			t.Errorf("DFA state name %q should be in {braces}", s)
		}
	}
}

// =========================================================================
// ToDot tests
// =========================================================================

func TestNFA_ToDot(t *testing.T) {
	nfa := makeContainsABNFA()
	dot := nfa.ToDot()

	checks := []string{
		"digraph NFA",
		"rankdir=LR",
		"__start",
		"doublecircle",
	}
	for _, check := range checks {
		if !strings.Contains(dot, check) {
			t.Errorf("ToDot() missing %q", check)
		}
	}
}

func TestNFA_ToDot_Epsilon(t *testing.T) {
	nfa := makeEpsilonNFA()
	dot := nfa.ToDot()

	// Should contain the epsilon character
	if !strings.Contains(dot, "\u03b5") {
		t.Error("ToDot() should contain epsilon character for epsilon transitions")
	}
}

// =========================================================================
// Edge cases
// =========================================================================

func TestNFA_SingleState_Accepting(t *testing.T) {
	nfa := NewNFA(
		[]string{"q0"},
		[]string{"a"},
		map[[2]string][]string{
			{"q0", "a"}: {"q0"},
		},
		"q0",
		[]string{"q0"},
	)

	if !nfa.Accepts([]string{}) {
		t.Error("single accepting NFA should accept empty")
	}
	if !nfa.Accepts([]string{"a"}) {
		t.Error("single accepting NFA should accept [a]")
	}
}

func TestNFA_DeadEnd(t *testing.T) {
	// NFA where processing leads to no states
	nfa := NewNFA(
		[]string{"q0", "q1"},
		[]string{"a", "b"},
		map[[2]string][]string{
			{"q0", "a"}: {"q1"},
			// q1 has no transitions at all
		},
		"q0",
		[]string{"q1"},
	)

	if !nfa.Accepts([]string{"a"}) {
		t.Error("should accept [a]")
	}
	if nfa.Accepts([]string{"a", "b"}) {
		t.Error("should reject [a, b] (q1 is a dead end)")
	}
}

func TestNFA_EpsilonToAccepting(t *testing.T) {
	// NFA where epsilon leads directly to accepting
	nfa := NewNFA(
		[]string{"q0", "q1"},
		[]string{"a"},
		map[[2]string][]string{
			{"q0", EPSILON}: {"q1"},
		},
		"q0",
		[]string{"q1"},
	)

	if !nfa.Accepts([]string{}) {
		t.Error("NFA with epsilon to accepting should accept empty")
	}
}
